src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,132 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final Object[] EMPTY_ELEMENT_DATA = {};
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] elementData;
+    private int size = 0;
+
+    public ArrayList() {
+        this.elementData = EMPTY_ELEMENT_DATA;
+    }
+
+    private void removeElementAtIndex(Object[] elements, int index) {
+        if ((size - 1) >= index) {
+            System.arraycopy(elements, index + 1, elements, index, (size - 1) - index);
+        }
+        elements[size -= 1] = null;
+    }
+
+    private void checkIndex(int index) {
+        if ((index > size - 1 || index < 0) || index == size) {
+            throw new ArrayListIndexOutOfBoundsException("index is out of bound of array");
+        }
+    }
+
+    private Object[] grow(int minCapacity) {
+        int oldCapacity = elementData.length;
+        if (oldCapacity > 0) {
+            int newCapacity = oldCapacity + (oldCapacity >> 1);
+            Object[] copyOfElementData = elementData;
+            elementData = new Object[Math.max(newCapacity,minCapacity)];
+            System.arraycopy(copyOfElementData,0, elementData, 0, oldCapacity);
+            return elementData;
+        }
+        return elementData = new Object[Math.max(DEFAULT_CAPACITY, minCapacity)];
+    }
```

Comment:
You should do it immediately on this array, without returning anything, for example like this:
```suggestion
    private void growIfArrayFull() {
        if (size >= arrayUnderneath.length) {
            int newArraySize = (int) Math.round(arrayUnderneath.length * GROWTH_MULTIPLIER);
            T[] newArray = (T[]) new Object[newArraySize];
            System.arraycopy(arrayUnderneath, 0, newArray, 0, arrayUnderneath.length);
            arrayUnderneath = newArray;
        }
    }
```